assets/src/edit-story/app/history/useHistoryReducer.js

```diff
@@ -42,9 +42,35 @@ const reducer = (size) => (state, { type, payload }) => {
         );
 
         if (isReplay) {
+          const currentEntry = state.entries[state.offset];
```

Comment:
Not very familiar with history system, so apologies if I'm wildly off base. But the bug describes the problem with UNDO, while the fix here is for the `ADD_ENTRY`?

Reply:
When replay is happening, `ADD_ENTRY` is triggered as well, which will then check if it's a replay or actually adding a new entry.

Dynamically assigning the page in `REPLAY` would give `false` for `isReplay` in `ADD_ENTRY` (as far as I understand) and create a new entry even if it shouldn't.

I guess that by naming doing all this in `REPLAY` is more logical when reading the code, I could def look into it to check if that would work out well.

Thoughts, @barklund ?

Reply:
I think this is very confusing. Ideally we would not trigger `ADD_ENTRY` in replay. E.g. maybe in our side effect we can skip it somehow and do all the work in the actual replying reducer?

I think that'd mean that [here](https://github.com/google/web-stories-wp/blob/47663f405f36f1466248db2341817645ef10c1fa/assets/src/edit-story/app/story/effects/useHistoryEntry.js#L40), we'd need add `isReplaying` boolean to the deps and call `appendToHistory()` only when `isReplaying === false`.

Reply:
Hmm, looks like currently the offset is also set in the `ADD_ENTRY` although I suppose it should be done in `REPLAY` then.

Reply:
So, the history provider is a little complex. A successful replay consists of two things happening in order:

* First a `REPLAY` request asking the history provider to tell you, what was in a previous (or future) history state relative to the current offset.
* Second, an `ADD_ENTRY` event with the **exact same state** as was returned by the `REPLAY` event (not returned though, but set in another object).

It is this object equality that is used to recognize, that replay has actually been effectuated in the history consumer and wasn't ignored.

The underlying idea is, that the history provider cannot force the consumer (here: the story provider) to go to a certain state, it can only ask it to try to, and then when the story provider (the history consumer) confirms this state is now the current state by calling `ADD_ENTRY`, only then does the history provider recognize, that the replay event was successful (by updating `offset`).

`ADD_ENTRY` could be renamed to `SET_CURRENT_STATE` or similar to clear up this minor confusion. `ADD_ENTRY` doesn't actually necessarily add an entry, it just says "this is my current state" and the history provider reacts accordingly (either confirming a successful replay or adding it as a new entry).

Reply:
@barklund: I'm personally more in favor of just changing the name (as also suggested below) instead of `useHistoryEntry` having to be aware and decide if to add an entry or clear the `replayState`. It feels like that shouldn't be a concern for `useHistoryEntry` which is basically just reacting to any relevant change.

Happy to change that part of the logic back and just adjust the name to be more accurate if everyone else is OK with it!

Thoughts, @dvoytenko?